### PR TITLE
Add `rsl::StaticVector` and `rsl::StaticString`

### PR DESCRIPTION
**Pull request by tylerjw**
- Link rclcpp and rcb_span
- Fixed size array and string

Migrated from https://github.com/PickNikRobotics/generate_parameter_library/blob/main/parameter_traits/include/parameter_traits/fixed_size_types.hpp

### CODE REVIEW

File: include/rsl/fixed_size.hpp [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+    /**
+     * @brief Allow default construction
+     */
+    FixedSizeString() = default;
+
+    /**
+     * @brief Construct from a std::string
+     */
+    FixedSizeString(std::string const& str) {
+        // on debug builds fail here with an assert
```
[the remark this answers is not in the document]
Comment (tylerjw) [reply include/rsl/fixed_size.hpp:145]: ok, I see your point here

File: include/rsl/fixed_size.hpp [hunk 2]
```diff
@@ -0,0 +1,212 @@
+#pragma once
```
[the remark this answers is not in the document]
Comment (tylerjw) [reply include/rsl/fixed_size.hpp:1]: doing that

File: include/rsl/fixed_size.hpp [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+     * @brief Construct from a std::string
+     */
+    FixedSizeString(std::string const& str) {
+        // on debug builds fail here with an assert
+        assert(str.size() <= Size);
+        // on release builds, continue on copying the most you can safely
+        len_ = std::min(str.size(), Size);
+        std::copy(
+            std::cbegin(str),
+            std::cbegin(str) +
```
[the remark this answers is not in the document]
Comment (tylerjw) [reply include/rsl/fixed_size.hpp:151]: I prefer the function version, converting them to the function version.

File: include/rsl/static_string.hpp [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+#include <type_traits>
+
+namespace rsl {
+
+/**
+ * @brief Fixed size string with conversion operators for std::string_view, std::string, and
+ * rclcpp::ParameterValue. At compile time you specify the maximum size of the container with
+ * the Size template parameter. At runtime you can populate up to Size values.
+ */
+template <size_t Size>
```
Comment (ChrisThrasher) [inline include/rsl/static_string.hpp:18]:
```suggestion
template <size_t capacity>
```

File: include/rsl/static_string.hpp [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+#include <array>
+#include <cassert>
+#include <string>
+#include <string_view>
+#include <type_traits>
+
+namespace rsl {
+
+/**
+ * @brief Fixed size string with conversion operators for std::string_view, std::string, and
```
Comment (ChrisThrasher) [inline include/rsl/static_string.hpp:14]:
```suggestion
 * @brief Fixed capacity string with conversion operators for std::string_view, std::string, and
```
Can we be more clear about where we mean size and where we mean capacity such that we match how the standard uses these terms?

File: include/rsl/static_vector.hpp [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+#include <tcb_span/span.hpp>
+
+#include <array>
+#include <cassert>
+#include <type_traits>
+#include <vector>
+
+namespace rsl {
+
+/**
```
Comment (ChrisThrasher) [inline include/rsl/static_vector.hpp:13]:
```suggestion
namespace rsl {

/** @file */

/**
```

Comment (ChrisThrasher) [reply include/rsl/static_vector.hpp:13]: Otherwise the file won't appear in the docs.

File: include/rsl/static_string.hpp [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+
+#include <array>
+#include <cassert>
+#include <string>
+#include <string_view>
+#include <type_traits>
+
+namespace rsl {
+
+/**
```
Comment (ChrisThrasher) [inline include/rsl/static_string.hpp:13]:
```suggestion
namespace rsl {

/** @file */

/**
```

File: include/rsl/static_vector.hpp [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+     */
+    template <typename Collection>
+    StaticVector(Collection const& collection) {
+        assert(collection.size() <= Size);
+        len_ = std::min(collection.size(), Size);
+        std::copy(
+            std::cbegin(collection),
+            std::cbegin(collection) +
+                static_cast<
+                    typename std::remove_reference_t<decltype(collection)>::difference_type>(len_),
```
Comment (ChrisThrasher) [inline include/rsl/static_vector.hpp:37]: What motivated you to add all this static casting and type trait code? Is adding the `size_t` `len_` not correct or error prone?

**Review by ChrisThrasher — CHANGES_REQUESTED** [review]

File: include/rsl/static_vector.hpp (hunk 8 above)
Comment (tylerjw) [reply include/rsl/static_vector.hpp:37]: Technically the difference_type inside a vector is a singed value, and Wconversion complains about adding a size_t to an iterator.

**Comment by ChrisThrasher** [comment]
I'm having a hard time accepting all the extra free functions and implicit conversations, but especially the extra free functions. Once the implicit conversions are in place, all we're doing is providing more verbose alternatives to existing standard constructors. I'd rather read
```cpp
auto sv = std::string_view(static_string);
```
than
```cpp
auto sv = rsl::to_string_view(static_string);
```
or even the slightly shorter
```cpp
auto sv = to_string_view(static_string);
```
if we want to rely on ADL. We're reinventing wheels. I do think some sort of extra convenient to convert from our static containers to standard containers is good, but I don't think this implicit conversion + free function combo is our best option.

---

I'd propose we add two more common container member functions in `.data()` and `.size()` so that our containers feel a bit more natural to those expecting an STL-like interface. This also makes our containers work with `std::data` and `std::size` overloads which makes them easier to use in generic code. Once we have those, we can simply ask users to convert as they want. Converting from owning container to owning container can be a heavyweight task so it's not unreasonable to require that users know what they're doing before gleefully copying a 10MB static vector into a `std::vector`.

I do think implicit conversions and/or free function _to non-owning containers_ is reasonable. These tasks are basically no-ops so it's perfectly suitable to make these operations really simple. In our case that means conversions to `std::string_view` and `tcb::span`.

**Review by ChrisThrasher — APPROVED** [review]

**Pull request merged**